File: gosub-bindings/src/nodes/text.c
#include "text.h"
#include "nodes.h"

const char *render_tree_node_text_value(const struct node_t *node) {
  if (!node)
    return NULL;

  return (const char *)node->data.text.value;
}

const char *render_tree_node_text_font(const struct node_t *node) {
  if (!node)
    return NULL;

  return (const char *)node->data.text.font;
}

float render_tree_node_text_font_size(const struct node_t *node) {
  if (!node)
    return 0.0f;

  return node->data.text.font_size;
}

bool render_tree_node_text_bold(const struct node_t *node) {
  if (!node)
    return false;

  return node->data.text.is_bold;
}

File: src/render_tree/text.rs
use std::{ffi::CString, os::raw::c_char};

#[derive(Debug, PartialEq, Clone, Copy)]
#[repr(C)]
pub struct TextNode {
    /// Body of the text node that will be drawn
    pub value: *mut c_char,
    pub font: *mut c_char,
    pub font_size: f32,
    pub is_bold: bool,
    // TODO: color, styles, visiblity, etc
}

impl TextNode {
    fn new(fs: f32, bold: bool) -> Self {
        Self {
            value: CString::new("").expect("").into_raw(),
            font: CString::new("Times New Roman").expect("").into_raw(),
            font_size: fs,
            is_bold: bold,
        }
    }

    pub fn new_heading1() -> Self {
        TextNode::new(37., true)
    }

    pub fn new_heading2() -> Self {
        TextNode::new(27.5, true)
    }

    pub fn new_heading3() -> Self {
        TextNode::new(21.5, true)
    }

    pub fn new_heading4() -> Self {
        TextNode::new(18.5, true)
    }

    pub fn new_heading5() -> Self {
        TextNode::new(15.5, true)
    }

    pub fn new_heading6() -> Self {
        TextNode::new(12., true)
    }

    pub fn new_paragraph() -> Self {
        TextNode::new(18.5, false)
    }
}
